[BUG] High Memory using Full Dashboard

### Component

Full Dashboard

### Performance Monitor Version

High Memory Usage and Not Releasing

### SQL Server Version

SQL Server 2019 CU32 GDR

### Windows Version

Windows Server 2019 Datacenter

### Describe the Bug

I am seeing high memory usage when using the Full Dashboard.  

<img width="508" height="88" alt="Image" src="https://github.com/user-attachments/assets/4ae36048-62a6-45fd-aa1c-6e06fb205d5a" />

It has gone as high as 4GB and doesn't seem to release without the process being killed.  Even when the dashboard is closed but still running in the background Memory continues to grow.  I am monitoring 1 staging environment which is a busy server.

I also notice when this happens the refresh button isn't available in the dashboard, but you can move between tabs and change the duration

### Steps to Reproduce

Open Dashboard (Run as Administrator)
Connect to a server
Run processes to simulate high CPU, locking etc
Watch Memory increase and not release

### Expected Behavior

I'd expect Memory to not grow that high and also release when not in use

### Actual Behavior

Memory fluctuates but stays high, randomly increasing.  Refresh Tab button is not available

<img width="628" height="81" alt="Image" src="https://github.com/user-attachments/assets/c89d75a3-8da8-44bf-8f78-7a84f5be19d5" />

### Error Messages / Log Output

```text
Log Output below
```

### Screenshots

<img width="654" height="469" alt="Image" src="https://github.com/user-attachments/assets/bfd72153-2377-46d9-9e4c-2a8d5b586199" />

<img width="966" height="804" alt="Image" src="https://github.com/user-attachments/assets/a0f3448e-692d-46a6-ae32-f5903cbf8791" />

### Additional Context

SQL Server agent is running, Developer Edition in use.  It is an Azure VM with SQL Server 2019 installed

I'm on the Overview tab
Auto refresh is set to off.  In settings it is set to 30 seconds
I only have one server installed

Log File Info

[2026-04-16 10:45:14.099] [ERROR] Error refreshing QueryPerformance data: Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
Exception: SqlException
Message: Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
StackTrace:    at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, SqlCommand command, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteReaderInternal(IAsyncResult asyncResult)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteReaderAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location ---
   at PerformanceMonitorDashboard.Services.DatabaseService.GetQueryStatsAsync(Int32 hoursBack, Nullable`1 fromDate, Nullable`1 toDate, Boolean fromSlicer) in D:\a\PerformanceMonitor\PerformanceMonitor\Dashboard\Services\DatabaseService.QueryPerformance.cs:line 1180
   at PerformanceMonitorDashboard.Services.DatabaseService.GetQueryStatsAsync(Int32 hoursBack, Nullable`1 fromDate, Nullable`1 toDate, Boolean fromSlicer) in D:\a\PerformanceMonitor\PerformanceMonitor\Dashboard\Services\DatabaseService.QueryPerformance.cs:line 1227
   at PerformanceMonitorDashboard.Controls.QueryPerformanceContent.RefreshAllDataAsync(Boolean fullRefresh) in D:\a\PerformanceMonitor\PerformanceMonitor\Dashboard\Controls\QueryPerformanceContent.xaml.cs:line 970
[2026-04-16 10:45:16.683] [INFO] === Application Exiting (Exit Code: 0) ===

[2026-04-16 11:43:11.157] [ERROR] Error refreshing QueryPerformance data: Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
Exception: SqlException
Message: Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
StackTrace:    at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, SqlCommand command, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteReaderInternal(IAsyncResult asyncResult)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteReaderAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location ---
   at PerformanceMonitorDashboard.Services.DatabaseService.GetQueryStatsAsync(Int32 hoursBack, Nullable`1 fromDate, Nullable`1 toDate, Boolean fromSlicer) in D:\a\PerformanceMonitor\PerformanceMonitor\Dashboard\Services\DatabaseService.QueryPerformance.cs:line 1180
   at PerformanceMonitorDashboard.Services.DatabaseService.GetQueryStatsAsync(Int32 hoursBack, Nullable`1 fromDate, Nullable`1 toDate, Boolean fromSlicer) in D:\a\PerformanceMonitor\PerformanceMonitor\Dashboard\Services\DatabaseService.QueryPerformance.cs:line 1227
   at PerformanceMonitorDashboard.Controls.QueryPerformanceContent.RefreshAllDataAsync(Boolean fullRefresh) in D:\a\PerformanceMonitor\PerformanceMonitor\Dashboard\Controls\QueryPerformanceContent.xaml.cs:line 970


The auto refresh is set to off

<img width="720" height="149" alt="Image" src="https://github.com/user-attachments/assets/ba9bd32a-22f4-4977-8dc6-a37eb67ee5ea" />


If I open the server and then close it immediately, the same thing happens, so it doesn't appear to be linked to the tab actually being open, the Memory grows regardless as long as it is has been opened.

--- a/Dashboard/ServerTab.xaml
+++ b/Dashboard/ServerTab.xaml
@@ -101,7 +101,7 @@
                 <StackPanel Grid.Row="0" Grid.Column="1" Grid.ColumnSpan="2" Orientation="Vertical" Margin="15,0,0,0">
                     <StackPanel Orientation="Horizontal">
                         <Button x:Name="ApplyToAllButton" Content="Apply to All Tabs" Click="ApplyToAllTabs_Click" Margin="2,0" Padding="12,5" Style="{DynamicResource AccentButton}"/>
-                        <Button x:Name="RefreshButton" Content="Refresh Tab" Click="RefreshButton_Click" Margin="2,0" Padding="12,5" Style="{DynamicResource SuccessButton}"/>
+                        <Button x:Name="RefreshButton" Content="Refresh Tab" Click="RefreshButton_Click" Margin="2,0" Padding="12,5" Style="{DynamicResource SuccessButton}" ToolTip="Refresh the current tab. Ctrl+Click to refresh all tabs."/>
                         <ToggleButton x:Name="AutoRefreshToggle" Content="Auto-Refresh: Off" Click="AutoRefreshToggle_Click" Margin="8,0,2,0" Padding="12,5" MinWidth="130" ToolTip="Toggle auto-refresh on/off. Configure interval in Settings."/>
                         <Button Content="Edit Schedules" Click="EditSchedules_Click" Margin="8,0,2,0" Padding="12,5" Style="{DynamicResource AccentButton}" ToolTip="Edit collector schedules (frequency, retention, enabled/disabled)"/>
                         <TextBlock Text="|" VerticalAlignment="Center" Margin="8,0,4,0" Foreground="{DynamicResource ForegroundBrush}"/>

--- a/Dashboard/ServerTab.xaml.cs
+++ b/Dashboard/ServerTab.xaml.cs
@@ -51,6 +51,7 @@ namespace PerformanceMonitorDashboard
         private bool _isRefreshing;
         private DateTime _refreshStartedUtc;
         private bool _suppressPickerUpdates;
+        private readonly HashSet<string> _initializedTabs = new();
 
         // Filter state dictionaries for each DataGrid
 
@@ -424,6 +425,7 @@ namespace PerformanceMonitorDashboard
             _autoRefreshCts?.Cancel();
             _autoRefreshTimer?.Stop();
             _autoRefreshTimer = null;
+            _initializedTabs.Clear();
 
             Helpers.ThemeManager.ThemeChanged -= OnThemeChanged;
             Loaded -= ServerTab_Loaded;
@@ -549,7 +551,8 @@ namespace PerformanceMonitorDashboard
                 if (e.Key == System.Windows.Input.Key.F5)
                 {
                     e.Handled = true;
-                    await LoadDataAsync();
+                    bool fullRefresh = System.Windows.Input.Keyboard.Modifiers.HasFlag(System.Windows.Input.ModifierKeys.Control);
+                    await LoadDataAsync(fullRefresh);
                 }
                 else if (e.Key == System.Windows.Input.Key.V &&
                          System.Windows.Input.Keyboard.Modifiers == System.Windows.Input.ModifierKeys.Control &&
@@ -640,7 +643,7 @@ namespace PerformanceMonitorDashboard
                 CriticalIssuesTab.SetTimeRange(_globalHoursBack, _globalFromDate, _globalToDate);
                 DefaultTraceTab.SetTimeRange(_globalHoursBack, _globalFromDate, _globalToDate);
 
-                await LoadDataAsync();
+                await LoadDataAsync(fullRefresh: false);
                 SetupAutoRefresh();
             }
             catch (Exception ex)
@@ -654,7 +657,8 @@ namespace PerformanceMonitorDashboard
         {
             try
             {
-                await LoadDataAsync();
+                bool fullRefresh = System.Windows.Input.Keyboard.Modifiers.HasFlag(System.Windows.Input.ModifierKeys.Control);
+                await LoadDataAsync(fullRefresh);
             }
             catch (Exception ex)
             {
@@ -1271,7 +1275,9 @@ namespace PerformanceMonitorDashboard
         }
 
         /// <summary>
-        /// Refreshes only the currently visible tab — used on auto-refresh timer tick.
+        /// Refreshes only the currently visible tab. On first visit to a tab,
+        /// does a full refresh so all sub-tabs are populated. Subsequent visits
+        /// only refresh the active sub-tab for speed.
         /// </summary>
         private async Task RefreshVisibleTabAsync()
         {
@@ -1279,6 +1285,7 @@ namespace PerformanceMonitorDashboard
             if (selectedTab == null) return;
 
             var tabHeader = GetTabHeaderText(selectedTab);
+            bool firstVisit = _initializedTabs.Add(tabHeader);
 
             switch (tabHeader)
             {
@@ -1286,19 +1293,19 @@ namespace PerformanceMonitorDashboard
                     await RefreshOverviewTabAsync();
                     break;
                 case "Queries":
-                    await RefreshQueriesTabAsync(fullRefresh: false);
+                    await RefreshQueriesTabAsync(fullRefresh: firstVisit);
                     break;
                 case "Resource Metrics":
-                    await RefreshResourceMetricsTabAsync(fullRefresh: false);
+                    await RefreshResourceMetricsTabAsync(fullRefresh: firstVisit);
                     break;
                 case "Memory":
-                    await RefreshMemoryTabAsync(fullRefresh: false);
+                    await RefreshMemoryTabAsync(fullRefresh: firstVisit);
                     break;
                 case "Locking":
                     await RefreshLockingTabAsync();
                     break;
                 case "System Events":
-                    await RefreshSystemEventsTabAsync(fullRefresh: false);
+                    await RefreshSystemEventsTabAsync(fullRefresh: firstVisit);
                     break;
                 // Plan Viewer has no data to refresh
             }

--- a/Dashboard/Services/DatabaseService.QueryPerformance.cs
+++ b/Dashboard/Services/DatabaseService.QueryPerformance.cs
@@ -1110,7 +1110,7 @@ ORDER BY bucket_hour;";
                 qs.query_hash,
                 qs.creation_time
         )
-        SELECT
+        SELECT TOP (500)
             database_name = pl.database_name,
             query_hash = CONVERT(nvarchar(20), pl.query_hash, 1),
             object_type = MAX(pl.object_type),
@@ -1345,7 +1345,7 @@ ORDER BY bucket_hour;";
                 ps.object_name,
                 ps.cached_time
         )
-        SELECT
+        SELECT TOP (500)
             database_name = pl.database_name,
             object_id = MAX(pl.object_id),
             object_name = QUOTENAME(pl.schema_name) + N'.' + QUOTENAME(pl.object_name),
@@ -1520,7 +1520,7 @@ ORDER BY bucket_hour;";
                     string query = @"
         SET TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
 
-        SELECT
+        SELECT TOP (500)
             database_name = qsd.database_name,
             query_id = qsd.query_id,
             execution_type_desc = MAX(qsd.execution_type_desc),

